module/mempool/stdmap/backend.go

```diff
@@ -215,23 +216,25 @@ func (b *Backend) RegisterEjectionCallbacks(callbacks ...mempool.OnEjection) {
 func (b *Backend) reduce() {
 
 	// we keep reducing the cache size until we are at limit again
-	for len(b.entities) > int(b.limit) {
+	// this was a loop, but the loop is now in EjectTrueRandomFast()
+	if len(b.entities) > int(b.limit) {
 
 		// get the key from the eject function
-		key, _ := b.eject(b.entities)
+		key, entity, ok := b.eject(b)
 
 		// if the key is not actually part of the map, use stupid fallback eject
-		entity, ok := b.entities[key]
 		if !ok {
-			key, _ = EjectFakeRandom(b.entities)
-		}
+			// this will eject multiple entities, return values are unused
+			_, _, _ = EjectTrueRandomFast(b)
+		} else {
 
-		// remove the key
-		delete(b.entities, key)
+			// remove the key
+			delete(b.entities, key)
 
-		// notify callback
-		for _, callback := range b.ejectionCallbacks {
-			callback(entity)
+			// notify callback
+			for _, callback := range b.ejectionCallbacks {
+				callback(entity)
+			}
 		}
```

Comment:
I find this interaction with the backend's internal logic a bit confusing: https://github.com/onflow/flow-go/blob/070c71e7861284613b2c204800b7865f39ef1be0/module/mempool/stdmap/backend.go#L222-L228

With the default configuration, we call https://github.com/onflow/flow-go/blob/070c71e7861284613b2c204800b7865f39ef1be0/module/mempool/stdmap/backend.go#L223 which is the `EjectTrueRandomFast` policy. Since this policy _always returns false_, we execute https://github.com/onflow/flow-go/blob/070c71e7861284613b2c204800b7865f39ef1be0/module/mempool/stdmap/backend.go#L227-L228 which is a no-op. 

I like your approach of handing the full `Backend` instance to the ejection policy: https://github.com/onflow/flow-go/blob/070c71e7861284613b2c204800b7865f39ef1be0/module/mempool/stdmap/eject.go#L17

Overall, I feel we don't need this fallback anymore:
* We have to assume correct functioning of code to some degree. Why not also assume correct functioning of the ejection policy here? 
* Sure, if the ejection policy is buggy, the mempool will grow without bounds. But falling back on a different ejection policy other than the configured one also has major risks: it could undermine liveness of the entire protocol.

My suggestion would be to replace this entire code block by:
```golang
	// we keep reducing the cache size until we are at limit again
	// this was a loop, but the loop is now in EjectTrueRandomFast()
	if len(b.entities) > int(b.limit) {
		b.eject(b)
	}
```
and make the ejection policies do the `b.ejectionCallbacks`. This would be a more consistent design in my opinion.

Reply:
I agree completely.  I was trying to make minimal changes to existing logic.  It appears that that logic doesn't really help here.  I'll make this change.

Reply:
This is changed to your suggestion in the most recent commit.